In [279]:
import numpy as np
import pandas as pd

In [280]:
import pandas as pd
filepath = r"C:\Users\ejiro\Downloads\autos.csv"
df = pd.read_csv(filepath, encoding="ISO-8859-1")

In [281]:
df.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

In [282]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In [216]:
df.shape

(371528, 20)

In [287]:
# create a new dataframe that we would be working on
autos = df.copy()

In [289]:
# check the columns
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [290]:
# rename all columns to lowercase
autos.columns = autos.columns.str.lower()

In [291]:
#rename specific columns
autos = autos.rename(columns={'datecrawled': 'date_crawled', 'yearofregistration': 'registration_year', 'abtest':'ab_test', 'vehicletype':'vehicle_type', 'monthofregistration': 'registration_month',
                              'notrepaireddamage':'unrepaired_damage', 'datecreated': 'ad_created', 'nrofpictures':'number_of_pictures',
                             'offertype':'offer_type', 'fueltype':'fuel_type', 'lastseen': 'last_seen', 'postalcode':'postal_code',
                             'gearbox':'gear_box', 'powerps':'ps_power'})

In [292]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'ps_power', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'number_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [293]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  registration_year gear_box       ps_power  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  number_of_pictures  \
count             299468               371528            371528.0   
unique                 2                  114                 NaN   
top                 nein  2016-04-03 00:00:00                 NaN   
freq              263182                14450                 NaN   
mean                 NaN                  NaN                 0.0   
std                  NaN                  NaN                 0.0   
min                  NaN                  NaN                 0.0   
25%                  NaN                  NaN                 0.0   
50%                  NaN                  NaN                 0.0   
75%                  NaN                  NaN                 0.0   
max                  NaN                  NaN                 0.0   

         postal_code            last_seen  
count   371528.00000               371528  
unique           NaN               182806  
top              NaN  2016-04-07 06:45:59  
freq             NaN                   17  
mean     50820.66764                  NaN  
std      25799.08247                  NaN  
min       1067.00000                  NaN  
25%      30459.00000                  NaN  
50%      49610.00000                  NaN  
75%      71546.00000           

In [296]:
# explore the dataset for null values and convert to percentages

autos.apply(pd.isnull).sum().sort_values(ascending = False)  / autos.shape[0]

unrepaired_damage     0.193956
vehicle_type          0.101928
fuel_type             0.089861
model                 0.055134
gear_box              0.054394
kilometer             0.000000
postal_code           0.000000
number_of_pictures    0.000000
ad_created            0.000000
brand                 0.000000
registration_month    0.000000
date_crawled          0.000000
name                  0.000000
ps_power              0.000000
registration_year     0.000000
ab_test               0.000000
price                 0.000000
offer_type            0.000000
seller                0.000000
last_seen             0.000000
dtype: float64

In [298]:
# create two dataframes to handle the non and non null values 
# Determine the null values:
null_values = autos.isnull().sum().to_frame()
null_values = null_values.rename(columns = {0:'null'})

# Determine the not null values:
not_null = autos.notna().sum().to_frame()
not_null = not_null.rename(columns = {0:'not null'})

# Combine the dataframes:

null_count = pd.concat([null_values, not_null], ignore_index=False, axis=1).reset_index()
null_count = null_count.rename(columns = {'index':'category'})

# Display the new dataframe:
null_count

category   null  not null
0         date_crawled      0    371528
1                 name      0    371528
2               seller      0    371528
3           offer_type      0    371528
4                price      0    371528
5              ab_test      0    371528
6         vehicle_type  37869    333659
7    registration_year      0    371528
8             gear_box  20209    351319
9             ps_power      0    371528
10               model  20484    351044
11           kilometer      0    371528
12  registration_month      0    371528
13           fuel_type  33386    338142
14               brand      0    371528
15   unrepaired_damage  72060    299468
16          ad_created      0    371528
17  number_of_pictures      0    371528
18         postal_code      0    371528
19           last_seen      0    371528

In [299]:
# drop null values from the columns with null values
autos.dropna(subset=['fuel_type', 'unrepaired_damage', 'gear_box', 'vehicle_type', 'model'], inplace=True)

In [300]:
# confirm if null values are gone
autos[['model', 'fuel_type', 'unrepaired_damage','gear_box']].isnull().sum()

model                0
fuel_type            0
unrepaired_damage    0
gear_box             0
dtype: int64

In [302]:
# explore the individual columns and drop the ones where almost all values are the same
autos['seller'].value_counts()

privat        260954
gewerblich         2
Name: seller, dtype: int64

In [303]:
# drop seller column as it doesn't have enough unique values
del autos['seller']

In [304]:
autos['offer_type'].value_counts()

Angebot    260952
Gesuch          4
Name: offer_type, dtype: int64

In [305]:
# drop offer_type column also
del autos['offer_type']

In [307]:
# explore price column
autos['price'].describe()

count    2.609560e+05
mean     8.208628e+03
std      3.433091e+05
min      0.000000e+00
25%      1.500000e+03
50%      3.850000e+03
75%      8.600000e+03
max      1.000000e+08
Name: price, dtype: float64

In [313]:
# Price column Outlier detection and deletion

# calculate IQR for price
q1 = autos['price'].quantile(0.25)
q3 = autos['price'].quantile(0.75)
iqr = q3 - q1

# Identify outliers
threshold = 1.5
outliers = autos[(autos['price'] < (q1 - threshold * iqr)) | (autos['price'] > (q3 + threshold * iqr))]

autos = autos.drop(outliers.index)


In [314]:
# explore the dates columns
# There are 5 columns that should represent date values
#- `date_crawled`: added by the crawler
# - `last_seen`: added by the crawler
#- `ad_created`: from the website
#- `registration_month`: from the website
#- `registration_year`: from the website

autos[['date_crawled','ad_created','last_seen']].head()

date_crawled           ad_created            last_seen
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21
5  2016-04-04 17:36:23  2016-04-04 00:00:00  2016-04-06 19:17:07
6  2016-04-01 20:48:51  2016-04-01 00:00:00  2016-04-05 18:18:39
7  2016-03-21 18:54:38  2016-03-21 00:00:00  2016-03-25 16:47:58

In [316]:
# check the datatype
autos[['date_crawled','ad_created','last_seen']].dtypes

date_crawled    object
ad_created      object
last_seen       object
dtype: object

In [317]:
# convert to datetime

autos['date_crawled'] = pd.to_datetime(autos['date_crawled'])
autos['ad_created'] = pd.to_datetime(autos['ad_created'])
autos['last_seen'] = pd.to_datetime(autos['last_seen'])

In [318]:
autos[['date_crawled','ad_created','last_seen']].dtypes

date_crawled    datetime64[ns]
ad_created      datetime64[ns]
last_seen       datetime64[ns]
dtype: object

In [248]:
# explore the registration year column
autos['registration_year'].describe()

count    243793.000000
mean       2002.865611
std           6.130691
min        1910.000000
25%        1999.000000
50%        2003.000000
75%        2007.000000
max        2018.000000
Name: registration_year, dtype: float64

In summary, the 'registration_year' column primarily contains registration years from around 1910 to 2018. 
The majority of the vehicles in the dataset were registered between the late 1990s and the early 2000s, with relatively
little variability around the mean registration year of approximately 2002.87. There are also some older vehicles with 
registration years dating back to 1910.

In [340]:
#it's natural to explore variations across different car brands. We can use aggregation to understand the brand column.
brand_dist = autos['brand'].value_counts()
brand_dist

Volkswagen       51428
Bmw              26884
Opel             26155
Mercedes_benz    23508
Audi             21194
Ford             17040
Renault          11619
Peugeot           7818
Fiat              6326
Seat              4887
Skoda             4375
Mazda             3944
Citroen           3655
Toyota            3609
Nissan            3483
Smart             3347
Hyundai           2723
Mini              2636
Volvo             2456
Mitsubishi        2035
Honda             1924
Kia               1873
Suzuki            1677
Alfa_romeo        1677
Chevrolet         1273
Chrysler          1009
Dacia              723
Subaru             529
Daihatsu           520
Porsche            511
Jeep               494
Saab               413
Land_rover         413
Jaguar             375
Daewoo             337
Lancia             308
Rover              258
Trabant            211
Lada               146
Name: brand, dtype: int64

In [337]:
# remove space in column values using strip() function
autos['brand'] = autos['brand'].apply(lambda x: x.strip())

# capitalize first letter in values
autos['brand'].str.capitalize()

3         Volkswagen
4              Skoda
5                Bmw
6            Peugeot
7         Volkswagen
             ...    
371518           Bmw
371520          Seat
371521          Opel
371524         Smart
371525    Volkswagen
Name: brand, Length: 243793, dtype: object

3         Volkswagen
4              Skoda
5                Bmw
6            Peugeot
7         Volkswagen
             ...    
371518           Bmw
371520          Seat
371521          Opel
371524         Smart
371525    Volkswagen
Name: brand, Length: 243793, dtype: object

In [341]:
# Decide which brands to aggregate by (e.g., top 5)
top_brands = 5
brands_agg = brand_dist.head(top_n).index
print(f"\nTop {top_n} Brands to Aggregate By:")
print(brands_agg)


Top 5 Brands to Aggregate By:
Index(['Volkswagen', 'Bmw', 'Opel', 'Mercedes_benz', 'Audi'], dtype='object')


In [342]:
# Step 2: Create an empty dictionary to hold the aggregate data
brand_means = {}

In [343]:
# Step 3: Loop over the selected brands and calculate the mean price
for brand in brands_agg:
    subset = autos[autos['brand'] == brand]
    mean_price = subset['price'].mean()
    brand_means[brand] = mean_price
    
    sorted_brand_means = dict(sorted(brand_means.items(), key=lambda item: item[1], reverse=True))


In [344]:
# Step 4: Print the dictionary of aggregate data
print("Aggregate Data:")
print(sorted_brand_means)

Aggregate Data:
{'Audi': 6889.9186090402945, 'Bmw': 6725.753831275108, 'Mercedes_benz': 6275.939254721797, 'Volkswagen': 5062.88714319048, 'Opel': 3211.578168610208}


Audi, BMW and Mercedes Benz are more expensive, while Ford and Opel are less expensive